_posts/en/newsletters/2019-05-14-newsletter.md

```diff
@@ -0,0 +1,485 @@
+
+    FIXME: table example of HTLC cost in current P2WSH and taproot for
+    each of the three possible outcomes
+
+    P2WSH
+      Output: 8+1+34
+      1.preimage, digest, A pubkey, B pubkey, OP_IF/ENDIF, Alice sig
+      2. digest, A pubkey, B pubkey, IF/ENDIF, Bob sig
+      3. N/A (use one of the above)
+
+    Taproot
+      Output: 8+1+35
+      1. preimage, digest, A pubkey, A sig; internal key; leaf hash
+      2. B pubkey, B sig; internal key; leaf hash
+      3. AB sig
+
+Because we chose to use a small tree in the above example, we will
+describe some things that would've been clearer with a larger example.
+First, if you have any unpaired nodes in your tree, your software will
+be responsible for pairing them with something, e.g a 32-byte random
+number.  Additionally, the maximum depth of a tree is 32 rows, allowing
+a tree to contain a maximum of around four billion scripts.  However,
+only one of those scripts would appear in any spending transaction and
+only 32 hashes directly related to the merkle tree would need to be
+generated to prove the script was part of the tree---this means that more
+complex scripts than our simple HTLC can gain much larger efficiency
+improvements than seen above (see an article focused on [MAST][] for
+more information).
+
+### Slight changes to scripting with Tapscript
+
+When using Bitcoin Script with Taproot, some of the rules have been
+changed, most notably:
+
+- **Unused and invalid opcodes redefined:** the bytes of opcodes that
+  were never assigned, or which were made invalid by Satoshi Nakamoto,
+  or which were created for upgrades and have not been used yet have all
+  been changed to have the semantics of an `OP_SUCCESS` opcode that
+  unconditionally renders any script containing that code valid.
+  Because future soft forks can only add new rules by making
+  previously-valid things invalid, this maximal validity
+  today will allow maximal flexibility in future soft forks.  Of course,
+  receivers get to choose what scripts they receive payments to, so no
+  one who likes their money will include an `OP_SUCCESSn` opcode in
+  their scripts until a soft fork has given it a new consensus-enforced
+  meaning.
+
+- **Schnorr signatures required:** ECDSA signatures will not be accepted
+  in Tapscript.
+
+- **New script-based multisig semantics:** the current
+  `OP_CHECKMULTISIG` and `OP_CHECKMULTISIGVERIFY` opcodes are not
+  available in Tapscript.  Instead, the existing single-sig
+  `OP_CHECKSIG` and `OP_CHECKSIGVERIFY` may be chained (see example
+  below); or a new `OP_CHECKSIGADD` opcode may be used to increment a
+  counter if a signature matches a specified public key (e.g. for
+  2-of-3, you need two `OP_CHECKSIGADD` calls to succeed).  This change
+  is made to allow for batch verification of multiple signatures, which
+  [bip-schnorr][] estimates can speed up verification by FIXMEx over
+  checking each signature independently.
+
```

Comment:
Batch validation speedup, for a block of 1k-10k Schnorr signatures, is shown to be 2.2-2.6x in the Schnorr BIP[1]. However, that data included pubkey decompression. If that is factored out, then a speedup of 3-4x is anticipated for the same # of transactions[2].

[1] https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki
[2] https://bitcoin.stackexchange.com/questions/85213/when-schnorr-signatures-are-part-of-bitcoin-will-it-be-possible-validate-each-bl/85215